### PR TITLE
Control flow diagrams

### PR DESCRIPTION
**Pull request by MultipleNewts**
Add the automatic generation of control flow diagrams for a given ir output. This functionality can be accessed using flags in the cli.

### CODE REVIEW

File: src/python/__main__.py [hunk 1]
```diff
@@ -19,13 +19,19 @@
...
-    cmd.exec_file(filename, **args)
+    cmd.exec_file(filename, cfpath, **args)  # NOTE: idk is this works
 elif args.get("compile"):
-    cmd.compile_file(filename, **args)
+    cmd.compile_file(filename, cfpath, **args)  # NOTE: idk is this works
 else:
     with open(filename) as f:
         src = f.read()
-    cmd.translate(src, **args)
+    cmd.translate(src, cfpath, **args)
```
Comment (aurorusidk) [inline src/python/__main__.py:37]: the path does not need to be explicitly passed: it is contained within `**args`

File: src/python/control_flows.py
Comment (aurorusidk) [inline src/python/control_flows.py:1]: if there is a save path present the diagram probably shouldn't also be shown

**Comment by MultipleNewts** [comment]
ive noticed a coupe bugs, I'll fix those

File: src/python/cmd.py [hunk 2]
```diff
@@ -17,22 +18,30 @@ def translate(src, **flags):
...
+    block = noder.block
+    iir = noder.make(ast)
     if flags.get("make_ir"):
         print(IRPrinter().to_string(iir))
         return iir
     tc = TypeChecker()
     tc.check(iir)
     if flags.get("typecheck"):
         print(IRPrinter().to_string(iir))
+        if flags.get("cf_diagram"):
```
Comment (aurorusidk) [inline src/python/cmd.py:31]: this shouldn't be inside the `--typecheck` flag. `--make-ir` also produces IRs that may warrant a cf diagram

**Review by aurorusidk — CHANGES_REQUESTED** [review]

File: src/python/control_flows.py
Comment (aurorusidk) [reply src/python/control_flows.py:1]: for me the displaying opens a browser in a different workspace which is not ideal. personally I doubt I would ever want the diagram to show during translation but I may want to look at it later and would just have it saved. we might need a third opinion on this.

Comment (ActuallyEnbeon) [reply src/python/control_flows.py:1]: I agree with @aurorusidk, if the diagram is being saved then it shouldn't be displayed during translation

Comment (MultipleNewts) [reply src/python/control_flows.py:1]: 3 options:
- if saving, it must open
- if saving, it wont open
- perhaps if cfpath is present, but cf-diagram is not - save but don't open

Comment (MultipleNewts) [reply src/python/control_flows.py:1]: how about that last option i suggested?
seems to satisfy both

Comment (MultipleNewts) [reply src/python/control_flows.py:1]: @aurorusidk @ActuallyEnbeon

Comment (aurorusidk) [reply src/python/control_flows.py:1]: We could add an option to show and save but I don't think adding yet another command line argument is any better than just saving it and opening the file manually.

Comment (aurorusidk) [reply src/python/control_flows.py:1]: > how about that last option i suggested? seems to satisfy both

in that case the name should be different. maybe `--cf-show`?

Comment (MultipleNewts) [reply src/python/control_flows.py:1]: what I'm suggesting is:
--cf-diagram is present if you want a diagram to be shown or not
--cfpath [path] is for if you want to save it

so both saves and shows; just path saves but doesn't show; just cf-diagram shows but doesn't save; neither doesn't save or show
makes sense?

Comment (MultipleNewts) [reply src/python/control_flows.py:1]: that doesn't require another option, it utilises the 2 already present

Comment (aurorusidk) [reply src/python/control_flows.py:1]: ok, but as I said I think in that case `--cf-diagram` should be changed to more clearly communicate the functionality (like `--cf-show`)

Comment (MultipleNewts) [reply src/python/control_flows.py:1]: I see. That is reasonable, I will implement as such

**Review by aurorusidk — APPROVED** [review]

**Comment by aurorusidk** [comment]
Resolves #11

**Pull request merged**